the-collab-lab/tcl-37-smart-shopping-list
# Create new shopping list

## Description

This PR generates a token, if a token is not in local storage, on the root route, via a "Create New List" button.

The user is then taken to the list view.

The token is generated through the npm package installed from npm: https://www.npmjs.com/package/@the-collab-lab/shopping-list-utils

A shopping list consists of a set of items associated with a user’s token. Creating a new list consists of the following sequence:
* Generate a new, unique token
* Save the token to localStorage
* Show the user the list view

If a user visits the homepage and already has a token stored in localStorage, a useEffect within App.js will add that token to state and display a link to the shopping list. 

NOTE: We may wish to automatically redirect existing users to the list view in the future, but felt displaying a link would be less jarring for the user experience at this time.


## Related Issue


Close #3 

## Acceptance Criteria
- [X] @the-collab-lab/shopping-list-utils has been added as a dependency to the project.
- [X] For users who do not already have a token or a list, a button or link exists on the home screen that allows them to create a new list

* For this story, it's sufficient just to have a simple "create list" button. Although you are not required to save the token to Firestore, some teams have opted to also save the token to Firestore when generating a token.


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|  ✓  | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

The token is accessible in the <Home /> component, passed as a prop from the App class.

```<Home token={token} setToken={setToken} />```<br/>

For next steps: We will need to take the token as a team and create a list by the name of the token in the Firestore database. The '/list' view should show that particular list, by it's token from the db. This will be the connection between Issue #3 and Issue #4.

### Before


No token in local storage: <br/>
![Screen Shot 2022-01-19 at 5 52 14 PM](https://user-images.githubusercontent.com/67841237/150231488-8650c448-190d-461e-94d2-7f9f238cb55a.png)


### After
After pressing the "Create new list" button, token is generated and saved to local storage, and user is routed to '/list' view. <br/>
![Screen Shot 2022-01-19 at 5 53 02 PM](https://user-images.githubusercontent.com/67841237/150231781-7607c05b-d338-4609-8a1b-d75fd932a4e1.png)




![Screen Shot 2022-01-19 at 5 53 24 PM](https://user-images.githubusercontent.com/67841237/150231788-5157a7cb-8587-4080-9f73-3078ed3426d0.png)
 

## Testing Steps / QA Criteria


1. ```npm install``` (We added a new package)
2.  ```npm start```
3. Navigate to localhost:3000/
4. Click "Get new list" 
5. Expect to see a token in local storage, which can be viewed in Application, Storage in DevTools, or in Console ```localStorage.token```
6. You should expect to be sent to the "/list' route.
